Edit rounds ui!

jfly opened the pull request:
Now that #1619 is finally merged (afer over 4 months =(), we now have a `PATCH  /competitions/:id/wcif/events` endpoint!

This PR introduces a React based UI to edit the events and rounds of a competition. @Laura-O was kind enough to help out with the design of the UI components, which you can see in [this google doc](https://docs.google.com/document/d/1mpLfcJ0e8o-MFKJpiHpMgBUXpB29_9PptgtR8A2WEDc/).

You can try this out on staging!
- Edit UI: https://staging.worldcubeassociation.org/competitions/BerkeleyWinter2017/events/edit
- View events UI: https://staging.worldcubeassociation.org/competitions/BerkeleyWinter2017/events/


This PR is meant to be the minimum functionality required to let users input rounds and round attributes. I will address non critical stuff like UI polishing and validations in a later PR. I am tracking ideas for improvement in issue #1916.

## TODO
- [x] Do not allow delegates to add or remove events to a confirmed competition
- [x] Figure out bullet warnings, or at least add them to the whitelist [here](https://github.com/jfly/worldcubeassociation.org/blob/6ad50b1c749d4140caca8f94d2781498dd7eea0a/WcaOnRails/config/environments/development.rb#L76-L78).

pedrosino commented:
Can't see the google doc link.

jfly commented:
Oops! Updated the original post.

coder13 commented:
I'm initially concerned with the selection of saying if it's being held / how many rounds. With it being grayed out, I expect not to be able to interact with it even though I'm supposed to interact with it.

Maybe for fun, not being able to select certain cutoffs when certain round formats are selected.
I.E, best of 3 cutoff when round format is bo1.

![image](https://user-images.githubusercontent.com/881394/29996220-613263e0-8faf-11e7-9cf0-4e5af008f44d.png)

jfly commented:
Thanks for the feedback, @coder13! I've fixed up the advancement condition UI. I've also added a feature spec of editing events/rounds, which required adding an annoying amount of polyfill junk just to make PhantomJS happy.

Laura-O commented:
Sorry for the delay, here is my feedback:

- The "To Advance" column is hidden for me:
<img width="330" alt="bildschirmfoto 2017-09-07 um 11 41 35" src="https://user-images.githubusercontent.com/362249/30157058-8b4352c8-93c1-11e7-9325-4367eda365bc.png">

- Updating/saving changes:
Seeing this as a whole interface, I think the "save"-button in the modals might be misleading. The schedule is only updated after clicking the "Update Competition"-button on the bottom of the page. Currently, the button changes the color after edits, but if you do not scroll to the bottom, you will not see the button.
I think an alert on the top of the page after a round was changed would help (something like "Rounds edited, click 'Update Competition' to update the schedule").

- The event panels are neat!
What do you think about letting the user add the individual panels manually?
When creating schedules with only a small number of events but many rounds, the empty panels look confusing to me.
Another alternative: I can't remember if there were any plans by the Board to let delegates select events when they create competitions. If there are, this data could be used to create the default panels (e.g., if the delegate selects 222-555 as events, the panels for 222-555 are displayed).

- The requirement to advance in percent should be limited to a maximum of 75% (it's even possible to enter >100%).

- Cutoffs can be set to 0 centiseconds.

- The reset button doesn't work when you have saved the conditions once and then open the modal again.

- The explanations in the modals for the time limit and the advance requirements are helpful (like "Competitors have 5 minutes for each of their solves."). Adding something similar to the cutoff modal would be nice.

- It's not possible to set limits to >60 minutes. This is relevant for cumulative time limits (e.g. 90 minutes for 5BLD).

jfly commented:
Thanks for the feedback, @Laura-O! Some responses:

> - The "To Advance" column is hidden for me:

Ah, you must be at an unfortunate screen size, then. If I make this look good for you (2 columns of events instead of 3), it will be wasted space for someone with my screen size. With some work, we can probably find something that's good for both of us, but for now, I don't think it's worth the effort. Let's address this later in #1916. Just to be clear, the page is still usable for you, right? There should be horizontal scrollbars when stuff doesn't fit, but I don't see them in your screenshot, which I'm hoping is because you're on a mac and they hide scrollbars, not because the scrollbars don't exist.

> I think an alert on the top of the page after a round was changed would help (something like "Rounds edited, click 'Update Competition' to update the schedule").

I like this idea! I've gone ahead and made this change in https://github.com/jfly/worldcubeassociation.org/commit/03def6c913fd0819c82049ab1f1f0544677e6023. I figured it would be more useful to just put the button in the alert, how does this look?

![image](https://user-images.githubusercontent.com/277474/30172773-94bbf4fc-93aa-11e7-969c-dbb0f59816d5.png)

> What do you think about letting the user add the individual panels manually?

Sorry, I'm not sure what you mean here. Maybe it would be best to chat in real time about this?

> When creating schedules with only a small number of events but many rounds, the empty panels look confusing to me.

So I tried setting the opacity of events that are not being held to 50% to try to reduce visual clutter, but I guess that's not good enough =(

> Another alternative: I can't remember if there were any plans by the Board to let delegates select events when they create competitions

Haha, that's not merely something we allow delegate to do, we require them to do it! Currently competitions are [required to have at least 1 event before being confirmed](https://github.com/jfly/worldcubeassociation.org/blob/03def6c913fd0819c82049ab1f1f0544677e6023/WcaOnRails/app/models/competition.rb#L150-L155).

> If there are, this data could be used to create the default panels (e.g., if the delegate selects 222-555 as events, the panels for 222-555 are displayed).

The new UI you see is an update of the current /events/edit page, so that page you're thinking of does not exist anymore. Just to be clear, here's what I currently see on https://www.worldcubeassociation.org/competitions/BerkeleyWinter2017/events/edit:

![image](https://user-images.githubusercontent.com/277474/30173047-494f11ec-93ab-11e7-97e2-2953007a8c69.png)

When we merge this PR, the above page will change to have all the event panels you've been playing around with.

> The requirement to advance in percent should be limited to a maximum of 75% (it's even possible to enter >100%).

Done in https://github.com/jfly/worldcubeassociation.org/commit/ab57c0b92ca4fe362411127b1762849aac354545.

> Cutoffs can be set to 0 centiseconds.

This is a little tricky to implement, and I didn't build the infrastructure for validations yet. I'm going to suggest that we address this in #1916.

> The reset button doesn't work when you have saved the conditions once and then open the modal again.

The reset button resets the modal back to the state it was in at the time you opened the modal dialog. It sounds like you expect it to reset the modal back to the state the WCIF was in when you first visited the page? I personally think the current behavior is more useful, because if you want to reset the WCIF back to the state it was in when you first visited the page, you can refresh the page. However if this is just too confusing, maybe we should get rid of the reset buttons entirely?

> The explanations in the modals for the time limit and the advance requirements are helpful (like "Competitors have 5 minutes for each of their solves."). Adding something similar to the cutoff modal would be nice.

Done in https://github.com/jfly/worldcubeassociation.org/commit/f67cf2ca08bcc890c0879e253feee8e1c0e9bed4. How does this look?

![image](https://user-images.githubusercontent.com/277474/30174343-9c8fdac2-93af-11e7-9084-71b67fcf766e.png)

> It's not possible to set limits to >60 minutes. This is relevant for cumulative time limits (e.g. 90 minutes for 5BLD).

Ah, good point! Fixed in https://github.com/jfly/worldcubeassociation.org/commit/abcd2544686294d526e0405e764ffe786f45b743.

jfly commented:
Updated in response to review comments. I pushed some non-critical suggested out into #1916, which I'll address in a later PR. For now, I believe this is ready to go, and I hope to merge this up on Monday (but I'd really love some people to take a look at the code first!)

coder13 commented:
Code looks good to me. :+1:

Laura-O commented:
Skipping some comments as I was confused and @jfly already enlightened me when we chatted about this. :smile:

> Ah, you must be at an unfortunate screen size, then. If I make this look good for you (2 columns of events instead of 3), it will be wasted space for someone with my screen size. With some work, we can probably find something that's good for both of us, but for now, I don't think it's worth the effort right now. Let's address this later in #1916. Just to be clear, the page is still usable for you, right? There should be horizontal scrollbars when stuff doesn't fit, but I don't see them in your screenshot, which I'm hoping is because you're on a mac and they hide scrollbars, not because the scrollbars don't exist.

Yes, the screenshot was on a 13" MBP. It's still usable.

> The reset button resets the modal back to the state it was in at the time you opened the modal dialog. It sounds like you expect it to reset the modal back to the state the WCIF was in when you first visited the page? I personally think the current behavior is more useful, because if you want to reset the WCIF back to the state it was in when you first visited the page, you can refresh the page. However if this is just too confusing, maybe we should get rid of the reset buttons entirely?

Yes, maybe removing is the best solution.
If there is a "reset"-button I expect it to have some functionality. I think showing a reset button, but not allowing resets is weird and I am sure users will report this as a bug.

> Done in jfly/worldcubeassociation.org@f67cf2c. How does this look?

Looks good! :thumbsup:

jfly commented:
> If there is a "reset"-button I expect it to have some functionality. I think showing a reset button, but not allowing resets is weird and I am sure users will report this as a bug.

(It's very possible I'm misunderstanding you here, so please tell me if I am.) I don't think it's fair to describe this as "not allowing resets". It seems to me that we just have different definitions of "reset". My definition is "reset takes you back to the state the modal was in when you opened the modal". Your definition is "reset takes you back to the state this was in when you first loaded the page". I'd like to get some other people's opinions on what they think makes sense here.

FatBoyXPC commented:
Reset should reset to when I first started interacted with whatever thing I am trying to reset.

coder13 commented:
It works exactly as I expect it to, does something that I want but I can see where Laura is coming from and maybe it's not the most properly named? Sometimes I do some changes and want to go back to where it was. I might close out of the dialog and reopen it, or hit the reset button.
Maybe add a "default" button? Or remove the Reset button? If anything at all.

jfly commented:
I had a chat with @FatBoyXPC and @coder13 and we agreed that the behavior of the dialogs was confusing. Here's the system we decided is best:

> i'm getting rid of the save button and the reset button
> we're going to have an OK and a Close button
> close, clicking X, and clicking in the black background will all ask you if you are sure you want to discard your changes before discarding your changes and closing the dialog. if you don't have any changes in the dialog, then you won't be prompted before the dialog is closed
> the OK button will "save" your changes, which doesn't mean writing to the database
> you still have to hit the save button on the main page to write to the db
> (we all agreed that currenty, the save button is very poorly named because it doesn't actually write to the db)

I've implemented this in dca0f2b669845945d9be2e57b0d4387b8a72f9dd, and you can see it live on staging now!

jonatanklosko commented on WcaOnRails/app/javascript/edit-events/modals/index.jsx:
```diff
@@ -0,0 +1,151 @@
+    });
+  });
+  return roundsSharingTimeLimit;
+}
```
What about this?
```js
function findRoundsSharingTimeLimitWithRound(wcifEvents, wcifRound) {
  return _.flatMap(wcifEvents, 'rounds').filter(otherWcifRound =>
    otherWcifRound !== wcifRound
    && otherWcifRound.timeLimit
    && otherWcifRound.timeLimit.cumulativeRoundIds.includes(wcifRound.id)
  );
}
```

jonatanklosko commented on WcaOnRails/app/javascript/edit-events/modals/index.jsx:
```diff
@@ -0,0 +1,151 @@
+    });
+  });
+  return wcifRounds;
+}
```
What about this?
```js
function findRounds(wcifEvents, roundIds) {
  return _.flatMap(wcifEvents, 'rounds').filter(wcifRound => roundIds.include(wcifRound.id));
}
```

jonatanklosko commented on WcaOnRails/app/javascript/edit-events/modals/index.jsx:
```diff
@@ -0,0 +1,151 @@
+  }
+
+  hasUnsavedChanges = () => {
+    return JSON.stringify(this.getSavedValue()) != JSON.stringify(this.state.value);
```
Let's Lodash do the trick `return _.isEqual(this.getSavedValue(), this.state.value);`

jfly replied:
Nit, we want to check if they're not equal, not if if they are equal =)

jonatanklosko commented on WcaOnRails/app/javascript/edit-events/modals/index.jsx:
```diff
@@ -0,0 +1,151 @@
+  }
+
+  onChange = (value) => {
+    this.setState({ value: value });
```
You can simply do `this.setState({ value });` as the key name matches the variable name =)

jonatanklosko commented on WcaOnRails/app/javascript/edit-events/modals/index.jsx:
```diff
@@ -0,0 +1,151 @@
+          throw new Error();
+        }
+        otherWcifRound.timeLimit.cumulativeRoundIds.splice(index, 1);
+      });
```
I think this could simply be:
```js
findRoundsSharingTimeLimitWithRound(this.props.wcifEvents, wcifRound).forEach(otherWcifRound => {
  _.pull(otherWcifRound.timeLimit.cumulativeRoundIds, wcifRound.id);
});
```

jonatanklosko commented on WcaOnRails/app/javascript/polyfills/index.js:
```diff
@@ -0,0 +1,50 @@
+import 'whatwg-fetch';
+import Promise from 'promise-polyfill';
```
I don't quite get why `Promise` and `#find` polyfills go here. Both are defined in `babel-polyfill` - that we already have (that should be imported here anyway, because I think it's not used at the moment).

jfly replied:
Ah, this stuff is here to make PhantomJS happy. I didn't know about `babel-polyfill`! I'll give it a shot and see if PhantomJS is happy.

jonatanklosko commented on WcaOnRails/app/models/competition.rb:
```diff
@@ -851,17 +851,30 @@ def set_wcif_events!(wcif_events)
+        event_to_be_added = wcif_event["rounds"] && !wcif_event["rounds"].empty?
+        if event_to_be_added
+          competition_events.find_by_event_id!(wcif_event["id"]).load_wcif!(wcif_event)
+        end
```
Just curious: why changing this behaviour?

jfly replied:
With the previous version of the code, there were situations where we would create and immediately delete a competition event.

Consider an incoming WCIF that has 3x3x3, but with 0 rounds. In the "# Create missing events." block above, we'd create a competition_event, and then in this "# Update all events" block, we'd call `load_wcif!` for that newly created competition_event, which would then delete the competition_event. This behavior was technically correct, but seemed wasteful to me, so I rewrote it as you see now.

jonatanklosko commented on WcaOnRails/config/environments/development.rb:
```diff
@@ -67,7 +67,7 @@
 
   config.after_initialize do
-    Bullet.enable = true
+    Bullet.enable = !ENVied.DISABLE_BULLET
```
Is there a scenario in which we want to disable Bullet in development?

jfly replied:
I need to look into this. When I first started working on this page, I was running into messages from Bullet that I could not fully track down. I introduced this environment variable so I wouldn't have to waste time dismissing the bullet dialog. I was worried I might be running into something like https://github.com/jfly/worldcubeassociation.org/blob/6ad50b1c749d4140caca8f94d2781498dd7eea0a/WcaOnRails/config/environments/development.rb#L76-L78.

jonatanklosko commented on WcaOnRails/spec/features/competition_events_spec.rb:
```diff
@@ -0,0 +1,104 @@
+  within(:css, ".panel.event-#{event_id}") do
+    yield
+  end
+end
```
I think you can use block forwarding to make these helpers more concise =)
```ruby
def within_event_panel(event_id, &block)
  within(:css, ".panel.event-#{event_id}", &block)
end
```

jonatanklosko commented on WcaOnRails/spec/features/competition_events_spec.rb:
```diff
@@ -0,0 +1,104 @@
+      save
+
+      expect(round_333_1.reload.time_limit).to eq TimeLimit.new(centiseconds: 5.minutes.in_centiseconds, cumulative_round_ids: [])
+      expect(round_333_1.reload.time_limit.to_s(round_333_1)).to eq "5:00.00"
```
1. The second `reload` is not needed.
2. We have a [`time_limit_to_s`](https://github.com/thewca/worldcubeassociation.org/blob/abc3d53cf39d186b31c3a0539a52fc794b59bb55/WcaOnRails/app/models/round.rb#L44-L46) method.

So
```ruby
round_333_1.time_limit_to_s
```

jonatanklosko commented on WcaOnRails/spec/features/competition_events_spec.rb:
```diff
@@ -0,0 +1,104 @@
+      end
+      save
+
+      expect(round_333_1.reload.cutoff.to_s(round_333_1)).to eq "2 attempts to get ≤ 2:00.00"
```
We have the neat [`cutoff_to_s`](https://github.com/thewca/worldcubeassociation.org/blob/abc3d53cf39d186b31c3a0539a52fc794b59bb55/WcaOnRails/app/models/round.rb#L48-L50) method, so `round_333_1.reload.cutoff_to_s`.

jonatanklosko commented on WcaOnRails/spec/features/competition_events_spec.rb:
```diff
@@ -0,0 +1,104 @@
+      end
+      save
+
+      expect(round_333_1.reload.advancement_condition.to_s(round_333_1)).to eq "Top 12 advance to round 2"
```
And the same thing applies here `round_333_1.reload.advancement_condition_to_s`.

jfly commented on WcaOnRails/app/javascript/wca/events.js.erb:
```diff
@@ -0,0 +1,26 @@
+
+export default {
+  official: <%= Event.official.to_json.html_safe %>.map(extend),
+  byId: mapObjectValues(<%= Event.all.index_by(&:id).to_json.html_safe %>, extend),
```
Nice, done!

jfly commented on WcaOnRails/app/javascript/edit-events/EditEvents.jsx:
```diff
@@ -0,0 +1,245 @@
+    this.setState({ saving: true });
+    promiseSaveWcif(wcif).then(response => {
+      return response.json().then(json => [ response, json ]);
+    }).then(([response, json]) => {
```
I like the use of `Promise.all` here!

Not sure if you also meant to make a point about indentation here, but I do not really like indenting my chained `then`s, because it looks weird when the chain ends and the level of indentation is more than 1 level off from the next one. See [here](https://github.com/beautify-web/js-beautify/issues/482#issuecomment-57046756) for a discussion including some folks who also have my school of thought on this.

I also don't want this PR to devolve into a discussion about javascript code formatting. I would like for us to get a linter and style checker set up, and then we can discuss how we want that thing to be configured. I've created #1924 to continue this discussion.

jfly commented on WcaOnRails/app/javascript/edit-events/EditEvents.jsx:
```diff
@@ -0,0 +1,245 @@
+      e.returnValue = confirmationMessage;
+      return confirmationMessage;
+    }
+  }
```
Ah, yikes, this looks hacky. Please see https://developer.mozilla.org/en-US/docs/Web/API/WindowEventHandlers/onbeforeunload. I've also updated the code to hopefully make it more clear what's going on here. It's kind of silly because there's this `confirmationMessage` that gets ignored by every browser.

jfly commented on WcaOnRails/app/javascript/edit-events/EditEvents.jsx:
```diff
@@ -0,0 +1,245 @@
+  );
+}
+
+const EventPanel = ({ wcifEvents, competitionConfirmed, wcifEvent }) => {
```
Good point. I was definitely learning a bit as I was going here. It would be cool if #1924 could enforce some rules about this for us.

jfly commented on WcaOnRails/app/javascript/edit-events/ButtonActivatedModal.jsx:
```diff
@@ -0,0 +1,79 @@
+import addEventListener from 'react-overlays/lib/utils/addEventListener';
+import ownerDocument from 'react-overlays/lib/utils/ownerDocument';
+
+export default class extends React.Component {
```
Good call. I did that here and over on `AttemptResultInput`.

jfly commented on WcaOnRails/app/javascript/edit-events/ButtonActivatedModal.jsx:
```diff
@@ -0,0 +1,79 @@
+    this.state = { showModal: false };
+  }
+
+  open = () => {
```
Ah! Because I wrote that code in `EditEvents` class before I knew you could define instance methods using the arrow function. I've changed the `EditEvents` class to get rid of the binds.

jonatanklosko replied:
Makes sense, to be honest I was surprised seeing this property assignment style, I was sure it's not a part of the ES6 spec, but we have a babel [plugin](https://github.com/jfly/worldcubeassociation.org/blob/6ad50b1c749d4140caca8f94d2781498dd7eea0a/WcaOnRails/.babelrc#L18) enabling this.

jfly commented on WcaOnRails/app/javascript/edit-events/ButtonActivatedModal.jsx:
```diff
@@ -0,0 +1,79 @@
+    }
+
+    this.props.reset();
+    this.setState({ showModal: false });
```
Ah, because I always default to writing my code by "bailing out" first. I'll flip this around as you suggest, because I know more people will find that normal.

jfly commented on WcaOnRails/app/javascript/edit-events/modals/AttemptResultInput.jsx:
```diff
@@ -0,0 +1,123 @@
+                 autoFocus={autoFocus}
+                 value={this.props.value}
+                 ref={c => this.movesInput = c}
+                 onChange={this.onChange} />
```
Good idea. I've added a min of 1 and a max of 80. I've also added a `min` of 1 for multiple blindfolded points.

jfly commented on WcaOnRails/app/javascript/edit-events/modals/Cutoff.jsx:
```diff
@@ -0,0 +1,93 @@
+      str = `Best of ${cutoff.numberOfAttempts} ${matchResult(cutoff.attemptResult, wcifEvent.id, { short: true })}`;
+    } else {
+      str = "-";
+    }
```
I decided to refactor things a bit and pulled this out into a `roundCutoffToString` function.

jfly commented on WcaOnRails/app/javascript/edit-events/modals/Cutoff.jsx:
```diff
@@ -0,0 +1,93 @@
+        newCutoff = null;
+      }
+      onChange(newCutoff);
+    };
```
Ah, so I'm not a big fan of multiline ternaries. I do like your idea to check for `=== 0` rather than `> 0` here, though, so I'm goiug to do that and then flip the `if...else` blocks.

jonatanklosko commented on WcaOnRails/app/javascript/edit-events/modals/TimeLimit.jsx:
```diff
@@ -0,0 +1,277 @@
+  }
+
+  hasUnsavedChanges = () => {
+    return JSON.stringify(this.props.selectedRoundIds) != JSON.stringify(this.getSelectedRoundIds());
```
*Self nit*: not equal.

jfly commented on WcaOnRails/app/javascript/edit-events/modals/utils.js:
```diff
@@ -0,0 +1,114 @@
+import events from 'wca/events.js.erb'
+
+function parseMbValue(mbValue) {
```
Ah yikes. Yes, I will do that, thank you for the reminder.

jfly commented on WcaOnRails/app/javascript/edit-events/modals/TimeLimit.jsx:
```diff
@@ -0,0 +1,277 @@
+        return;
+      }
+      wcifRounds = wcifRounds.concat(otherWcifEvent.rounds.filter(r => r != excludeRound));
+    });
```
I implemented something inspired by this use of `flatMap`. Please take a look at it after I commit my changes.

> I don't see the reason for rounds.filter(r => r != excludeRound), because when otherWcifEvent is the excluded round's event (excludeEventId), then we don't continue.

Good point! I definitely wrote this in the other order (exclude round, and *then* exclude event, and I didn't get rid of the excludeRound stuff). I'll change `excludeRound` to `excludeEventId`, which should hopefully be a little cleaner.

jonatanklosko replied:
It's great =)

jfly commented on WcaOnRails/app/javascript/edit-events/modals/TimeLimit.jsx:
```diff
@@ -0,0 +1,277 @@
+        case "cumulative":
+          cumulativeRoundIds = [wcifRound.id];
+          if(roundsSelector) {
+            cumulativeRoundIds = _.uniq(cumulativeRoundIds.concat(roundsSelector.getSelectedRoundIds()));
```
Oh yikes. The problem was that we'd end up with 2 entries of `wcifRound.id`. I've rewritten this as:

```ruby
cumulativeRoundIds = roundsSelector ? roundsSelector.getSelectedRoundIds() : [wcifRound.id];
```

jonatanklosko commented on WcaOnRails/app/javascript/edit-events/modals/Cutoff.jsx:
```diff
@@ -0,0 +1,104 @@
+function parseRoundId(roundId) {
+  let [eventId, roundNumber] = roundId.split("-");
+  return { eventId, roundNumber };
+}
```
There are three other places in this diff using `split("-")`, let's make use of the new function! =)

jonatanklosko commented on WcaOnRails/app/models/event.rb:
```diff
@@ -54,4 +54,20 @@ def fewest_moves?
+      canChangeTimeLimit: self.can_change_time_limit?,
+      isTimedEvent: self.timed_event?,
+      isFewestMoves: self.fewest_moves?,
+      isMultipleBlindfolded: self.multiple_blindfolded?,
```
Why do you prefer having camelCase on the ruby side instead of transforming the keys as suggested [here](https://github.com/thewca/worldcubeassociation.org/pull/1856/files/6ad50b1c749d4140caca8f94d2781498dd7eea0a..e7e08e45fcd9121a3e8868d24ff0b6e66e2b4410#r137937784)?

jfly replied:
Ah! I did start out with the camelCase transformation as you suggested, but then I realized for full javascriptiness, it's nicer to have `event.isTimedEvent` rather than `event.timedEvent`. Alternatively, we could do the camelcasing in javascript and live without `is*` prefixes, or we could have some heuristic that prefixes boolean values with "is", unless they already start with "is" or "can". What do you think is best?

jonatanklosko replied:
So I don't see a big problem with using the same name (and other case) in both worlds, so `is_multiple_blinfolded` looks ok as ruby parameter (unless Rubocop things otherwise?), and if `event.timed_event` is meaningful enough in ruby, why wouldn't `event.timedEvent` be in JS?

jfly replied:
Ah, I like you sugggestion to put the `is_` over in Ruby land. I would argue that `event.timed_event` is not very meaningful in ruby, whereas `event.timed_event?` *is*. Unfortunately, we don't have the suffix with "?" trick in Javascript land.

jfly replied:
Also, unfortunately, `_.mapKeys` receives it's arguments as `value, key`, so I had to rework things a bit. Please take a look after my next commit and let me know what you think!

jonatanklosko replied:
Ah, that's right, sorry for the invalid line I suggested. Currently it looks great!

jfly commented:
Thanks for the detailed, review, @jonatanklosko! I learned a lot about lodash =)

I'm going to wait for @viroulep to take a look before I merge this up.

viroulep commented:
I've talked offline to @jfly about two high level comments:
 - the greyed button @coder13 mentioned earlier
 - the layout of the show events

Jeremy just pushed to staging what is I believe an improvement to the UI, and makes it way more explicit how to "add" an event.
See this screenshot: 
![image](https://user-images.githubusercontent.com/1007485/30348274-1cf9ea78-980f-11e7-833b-4a29af34af05.png) (or test it live on staging on an unconfirmed competition such as [blabla2017](https://staging.worldcubeassociation.org/competitions/blabla2017/events/edit))

I reworked the layout of the show events page and the text for cutoff/time limit, here is a screenshot:
![show_events](https://user-images.githubusercontent.com/1007485/30348353-4e9bd398-980f-11e7-836c-a0cd498c3b2a.png).
There are more opportunities for improvements (such as i18n), but since the page won't be linked publicly they will be addressed later.
@jfly feel free to cherry-pick this: https://github.com/viroulep/worldcubeassociation.org/commit/cfde2c1ddb692cf55da804be3e9b70bc38212914

The last comment I have is the following: right now we cannot adjust the number of rounds for events of competitions which were confirmed before this PR (for example go to [WC2017](https://staging.worldcubeassociation.org/competitions/WC2017/events/edit)). I think we should be able to do so, because even if the show events page is not ready yet, it will probably be ready before some of the current announced competitions are over, and organizers may want to update the competition events rounds to display their info right away.


The overall work and features are absolutely awesome, thank a lot Jeremy! I'm 100% sure this will greatly help organizers and delegates, and I'd be glad to add a schedule generating thing on top of it :p.

jfly commented:
> @jfly feel free to cherry-pick this: viroulep/worldcubeassociation.org@cfde2c1

Done, thanks!

> right now we cannot adjust the number of rounds for events of competitions which were confirmed before this PR (for example go to WC2017).

Ah, excellent point! I decided the cleanest way to fix this was to give a different meaning to events with no rounds. See 0e79da914f57a7c99afa3bc534a7b054055f755f and check out https://staging.worldcubeassociation.org/competitions/WC2017/events/edit, which should be a functional UI now =)

viroulep commented on WcaOnRails/spec/models/competition_wcif_spec.rb:
```diff
@@ -115,12 +115,22 @@
+      expect(competition.events.map(&:id)).to match_array %w(333 333fm 333mbf 444)
+    end
+
+    it "does not remove competition event when wcif rounds are nil" do
```
nit: "it *does* remove competition event"?

jfly replied:
Doh!

viroulep commented:
Btw this is a major UI change, I think it would be really worth to send a mail to announcements to explain this change!

jfly commented:
Good idea, @viroulep! I've started drafting an announcement email [here](https://docs.google.com/document/d/1fgX4zZvqLXh2QLlg57Dgv01ZihSivKKYqMWmr6iU-R8/), comments are appreciated!

If there are no objections, I plan to merge this up on Friday the 15th.

lgarron commented:
> The requirement to advance in percent should be limited to a maximum of 75% (it's even possible to enter >100%).
> Done in jfly/worldcubeassociation.org@ab57c0b.

It seems that the [demo competition](https://staging.worldcubeassociation.org/competitions/BerkeleyWinter2017/events/edit) still has 100% set for the first round of multi. ;-)

jfly commented:
Yeah, it's just a silly ui input max right now. Later improvement will be to validate this server side.
